prov/shm/src/smr_dsa.c

```diff
@@ -185,23 +196,94 @@ static int dsa_open_wq(struct accfg_wq *wq)
 	return fd;
 }
 
-static void *dsa_idxd_wq_mmap(struct accfg_wq *wq)
+static bool dsa_write_is_supported(int fd) 
+{
+	int ret;
+	int max_comp_checks = 10000; // TODO: Should we use a different value? 
+	struct dsa_hw_desc desc = {0};
+	struct dsa_completion_record comp __attribute__((aligned(32)));
+	memset(&comp, 0, sizeof(comp));
+	
+	desc.opcode = DSA_OPCODE_NOOP;
+	desc.flags = IDXD_OP_FLAG_CRAV | IDXD_OP_FLAG_RCR;
+	desc.completion_addr = (unsigned long) &comp;
+
+	// make sure writes (e.g., comp.status = 0) are ordered wrt to write()
+	{ asm volatile("sfence":::"memory"); }
+
+	ret = write(fd, &desc, sizeof(desc));
+	
+	// TODO: Should we try more than once if write call fails?
+	if (ret != sizeof(desc))
+		return false;
+
+	// TODO: Do we add a pause like call? 
+	while (comp.status == 0 && max_comp_checks > 0) {
+		max_comp_checks--;
+		// _mm_pause();
+	}
+
+	if (comp.status == DSA_COMP_SUCCESS)
+		return true;
+
+	return false;
+}
+
+static int dsa_idxd_wq_mmap(struct accfg_wq *wq, bool *wq_mmap_support,
+			      bool *wq_write_support,
+			      bool *dsa_submit_cmd_mode_checked, void **wq_reg,
+			      int *wq_fd)
 {
 	int fd;
-	void *wq_reg;
+	int ret = -1;
+	void *mmap_ret = NULL;
 
 	fd = dsa_open_wq(wq);
 
-	wq_reg = mmap(NULL, LIMITED_MSIX_PORTAL_OFFSET, PROT_WRITE,
-		      MAP_SHARED | MAP_POPULATE, fd, 0);
-	close(fd);
-
-	if (wq_reg == MAP_FAILED) {
-		FI_WARN(&smr_prov, FI_LOG_EP_CTRL, "d_idxd_wq_mmap error\n");
-		return NULL;
+	if (fd < 0)
+		return ret;
+
+	if (*wq_mmap_support == true || *dsa_submit_cmd_mode_checked == false) {
```

Comment:
This logic seems convoluted. Keep the existing function the same and handle opening the fd for write seperately in the parent function.